src/libcore/str/mod.rs

```diff
@@ -2168,7 +2168,9 @@ impl str {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline(always)]
-    #[rustc_const_unstable(feature="const_str_as_bytes")]
+    #[cfg_attr(boostrap_stdarch_ignore_this, rustc_const_unstable(feature = "const_str_as_bytes"))]
+    // SAFETY: const safe because we transmute two types with the same layout
+    #[cfg_attr(not(boostrap_stdarch_ignore_this), allow_internal_unstable(const_fn_union))]
     pub const fn as_bytes(&self) -> &[u8] {
         union Slices<'a> {
```

Comment:
This union should be `repr(C)`, or one should at least add a comment here that says that this relies on unspecified behavior, because the offsets of `repr(Rust)` union fields are not guaranteed to be `0`, or even equal, but that there is a test somewhere in the compiler that checks for this particular case and would break if this changes.

Reply:
> just make transmute a const fn so that people stop using unions for transmute making this error all over the place.

I guess we could allow it in the same contexts where we allow union field access (`const` items, but not `const fn` bodies). @oli-obk that shouldn't really be able to introduce any new problems, should it?

Reply:
No, that would work

Reply:
I opened a PR for stabilizing const transmute outside of `const fn`: https://github.com/rust-lang/rust/pull/64011